src/recipes.jl

```diff
@@ -274,10 +274,19 @@ function Makie.plot!(gp::GraphPlot)
             if $(gp.elabels_rotation) isa Real
                 # fix rotation to a single angle
                 rot = $(gp.elabels_rotation)
+            elseif $(gp.elabels_rotation) == automatic
+                #point the labels up
+                rot = broadcast($to_angle, edge_paths[], $positions, gp.elabels_shift[])
+                upsidedownlabels = (rot .> π/2) .| (rot .< -π/2)
+                rot[upsidedownlabels] .+= π
+                gp.elabels_align[] isa Vector || (gp.elabels_align[] = fill(gp.elabels_align[], ne(gp.graph[]) ))
```

Comment:
I am not entirely happy with this because it might overwrite the user specified align. What do you think about defaulting the align to `(:center, automatic)`?
- `rot == automatic && align[2] == automatic` : add the `π` and chose top/bottom automatic, default behaviour
- `rot == automatic && align[2] != automatic`  : do not add the `π`, basicially same behaviour as before the pr, automatic rotation just aligns the text with the edge
- `rot != automatic && align[2] == automatic` : the rotation is specifially given, probably default `align[2]` to `:bottom`?
- `rot != automatic && align[2] != automatic` : well its all user specified

then it would be neccessary to store those in a new internal observable `elabels_align_internal` or something. Otherwise, the `automatic` arguments won't be persistent, ie if one for example drags nodes around, everything explicit derived from automatic needs to be derived again

Reply:
>  this because it might overwrite the user specified align

The purpose is to exactly to do not overwrite it. Rotating the uspide down labels about `π` basicly has as a result to show `:top` as `:bottom` and `:bottom` as `:top` (exactly as opposite). That's why later I switch this 2 labels in case it was reverted.
This has as a result to hold the specified align of the user.
E.g.
```julia
gc = circular_ladder_graph(5)
labs_align = insert!(fill((:center, :bottom), ne(gc)-1), 12, (:center, :top))
labs_cols = insert!(fill(:black, ne(gc)-1), 12, :red)
graphplot(gc, elabels = repr.(edges(gc)), nlabels=repr.(vertices(gc)), elabels_rotation=nothing, elabels_align=labs_align, elabels_color=labs_cols)
```
The label of the (Edge 6=>10) is on `:top` as specified from the user
![image](https://user-images.githubusercontent.com/27723218/156649432-0c3cc76c-d837-4366-baee-68678d24989b.png)
```
graphplot(gc, elabels = repr.(edges(gc)), nlabels=repr.(vertices(gc)), elabels_rotation=Makie.automatic, elabels_align=labs_align, elabels_color=labs_cols)
```
and stays on `:top` as specified by the user, even if the labels are rotated.
![image](https://user-images.githubusercontent.com/27723218/156649910-30614e4d-d996-4d0f-8ad5-904f4985c142.png)

So I think overall the user preferences are being preserved. Does something make you think different ?

> then it would be neccessary to store those in a new internal observable

Reply:
About the `elabels_opposite` I was thinking about its usability. In the end it is just a specific case of `elabels_rotation`. So do you think maybe it can be handled as an input to `elabels_rotation` e.g. a symbol of `:opposite` ?
If an element-wise operation is being followed as you suggested [above](https://github.com/JuliaPlots/GraphMakie.jl/pull/67#discussion_r818534834) you can end up having `elabels_rotation = (45, automatic, :opposite, ... )`

Reply:
> The way I think about the align is like the anchor of the text node (in the text node coordiante system). In the example above I'd clearly say that the anchor of the Edge 6 => 10 label is south, therefore (:center, :bottom), which is the opposite of the behaviour defined by the user.

I tested this on the master branch and the behavior is exactly the same. Maybe the reference system for Edge 6 => 10 is going from 6 to 10 with the left hand and then it's not south but north i.e. :top as specified from the user. As you also mention:

> edge labels are placed on the left side of the edge

Reply:
On master, changing between top and bottom changes the side of the label (from left to right) because the calculated angle stays the same. I think this is not good, because with automatic curved edges and stuff it is a good feature that edge labels are placed on one side per default.

My thinking is that the interface might be more straight forward for the user, if we explicitly specify the side of the label. Then, `:top` and `:bottom` just changes, whether the bottom of the glyphs is attached to the edge or the top. Defaulting this to `automatic` allows the recipe to choose :top or :bottom in way, that the text is upright.
The rotation of each edge label would have to be adjusted based on the combination of valign and side of the edge.

This way, if the label is in the way of another edge, the user would have to change `elabels_side = Dict(4=>:right)` to move the label to the other side. But the text would still be upright.

Reply:
Okey. I could push another commit with the following modifications:
- Use `getattr` for the edge label rotation part. (now behavior will be element-wise). Eg user could input something like [automatic, 45, 90, automatic]
- Introduce `elabels_side` to choose side of the edge label. (`:left` or `right` with `left` beigin the default)
- Introduce `elabels_anchor` in place of `elabels_align`
- As long as `elabels_anchor` is of `:x, automatic` Makie will show edge labels point up
- Make `elabels_opposite` obselete

These are breaking changes. Do we want to do the pull request from another branch ?

Future possible work
- Use `getattr` for all the recipe
- Use of Dict(Edge, Any)

Let me know if the above are fine or if I missed something.

Reply:
For me, breaking change would be fine. I think this set of changes would be quite the improvement for the label placement overall. I guess the most breaking/unexpected thing would be the renaming of align to anchor. I'm not entirely sure about that. If done, i think this change should be also done for the node labels before releasing.

Something which isn't explicitly in your list: I'd use `getattr` for `elabels_side` and `elabels_anchor` too.

> Do we want to do the pull request from another branch ?

I can try to add you to the repo, then you could work from a branch within. First I need to get the rights for that though.

I am grateful for your input and contribution. Please don't feel obliged to implement more stuff than you originally intended just because I said it'd be a nice idea! And please express clearly if you think this proposed interface isn't good or to far from what you envisioned. I'm not experienced with open source colab at all. Also feel free to contact me on Julia Slack or Zulip for further discussion if you happen to prefer this over Github.